staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go

```diff
@@ -384,6 +403,258 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 	}
 }
 
+func TestRoundTripSocks5AndNewConnection(t *testing.T) {
+	for _, redirect := range []bool{false, true} {
+		t.Run(fmt.Sprintf("redirect = %t", redirect), func(t *testing.T) {
+			localhostPool := x509.NewCertPool()
+			if !localhostPool.AppendCertsFromPEM(localhostCert) {
+				t.Errorf("error setting up localhostCert pool")
+			}
+
+			socks5Server := func(creds *socks5.StaticCredentials, interceptor *Interceptor) *socks5.Server {
+				var conf *socks5.Config
+				if creds != nil {
+					authenticator := socks5.UserPassAuthenticator{Credentials: creds}
+					conf = &socks5.Config{
+						AuthMethods: []socks5.Authenticator{authenticator},
+						Rewriter:    interceptor,
+					}
+				} else {
+					conf = &socks5.Config{Rewriter: interceptor}
+				}
+
+				ts, err := socks5.New(conf)
+				if err != nil {
+					t.Errorf("failed to create sock5 server: %v", err)
+				}
+				return ts
+			}
+
+			testCases := map[string]struct {
+				proxyAuth              *url.Userinfo
+				serverConnectionHeader string
+				serverUpgradeHeader    string
+				serverStatusCode       int
+				shouldError            bool
+			}{
+				"proxied tcp->tcp via SOCKS5": {
+					serverConnectionHeader: "Upgrade",
+					serverStatusCode:       http.StatusSwitchingProtocols,
+					shouldError:            false,
+				},
+				"proxied tcp->tcp via SOCKS5 with invalid auth": {
+					proxyAuth:              url.UserPassword("invalid", "auth"),
+					serverConnectionHeader: "Upgrade",
+					serverStatusCode:       http.StatusSwitchingProtocols,
+					shouldError:            true,
+				},
+				"proxied tcp->tcp via SOCKS5 with auth": {
+					proxyAuth:              url.UserPassword("proxyuser", "proxypasswd"),
+					serverConnectionHeader: "Upgrade",
+					serverStatusCode:       http.StatusSwitchingProtocols,
+					shouldError:            false,
+				},
+			}
+
+			for _, k := range sets.StringKeySet(testCases).List() {
+				testCase := testCases[k]
+				t.Run(k, func(t *testing.T) {
+					server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
```

Comment:
nit: can we extract this SPDY server construction into a standalone function to reduce code duplicate? Both SOCK5 tests and the existing tests can call this server construction function.

Reply:
Done!